aten/src/ATen/native/Indexing.cpp

```diff
@@ -226,34 +231,192 @@ static std::tuple<Tensor, Tensor> makeLinearIndex(Tensor self, TensorList orig)
   return std::make_tuple(self, linearIndex);
 }
 
-Tensor index(const Tensor & self, TensorList indices) {
-  AT_CHECK(indices.size() <= (size_t)self.dim(),
-           "too many indices for tensor of dimension ", self.dim(), " (got ", indices.size(), ")");
+static bool all_strides_match(TensorList tensors) {
+  AT_ASSERT(tensors.size() >= 1);
+  auto strides = tensors[0].strides();
+  for (auto& tensor : tensors.slice(1)) {
+    if (!strides.equals(tensor.strides())) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static std::string shapes_as_str(TensorList tensors) {
+  std::ostringstream os;
+  bool first = true;
+  for (auto& tensor : tensors) {
+    if (tensor.defined()) {
+      if (!first) {
+        os << ", ";
+      }
+      os << tensor.sizes();
+      first = false;
+    }
+  }
+  return os.str();
+}
+
+struct AdvancedIndex {
+  AdvancedIndex(const Tensor& src, TensorList indices);
+
+  Tensor src;
+  std::vector<Tensor> indices;
+  DimVector indexed_sizes;
+  DimVector indexed_strides;
+  int64_t dims_before;
+  int64_t dims_after;
+};
+
+static Tensor restride_src(const Tensor& src, int64_t dims_before, int64_t dims_indexed,
```

Comment:
If i understand correctly, this makes the src have the output size and 0 stride for the indexed dimensions. I can't find the reason why? How is such a Tensor useful?

Reply:
I believe this is a trick to be able to use the iterator approach: this way we can increase the counters for the indexers and the input / output tensors. The machinery for TensorIterator allow us to iterate over all elements of a number of tensors. But for indexing, we need to jump over arbitrary positions in the src tensor. By having this zero stride, we don't actually increase the indices, which enables us to manually jump over the elements given by the index

Reply:
Yes. For say `dst = self[idx1, idx2]`, it treats indexing like an element-wise operation over dst, self, and the indexers idx1 and idx2. After broadcasting and restriding the operation looks like:

`dst = *(self_ptr + (idx1 * stride(0)) + (idx2 * stride(1))`

where stride(0) is self.stride(0)

Unlike true element-wise operations, the value at self isn't meaningful, only the pointer to it. When you add the self pointer to the indexed values multiplied by the stride of the indexed dimension, you get a pointer to the true self value.